src/lib/homestore_backend/tests/hs_blob_tests.cpp

```diff
@@ -285,4 +287,135 @@ TEST_F(HomeObjectFixture, PGBlobIterator) {
     ASSERT_TRUE(pg_iter->update_cursor(objId(LAST_OBJ_ID)));
 }
 
-TEST_F(HomeObjectFixture, SnapshotReceiveHandler) {}
+TEST_F(HomeObjectFixture, SnapshotReceiveHandler) {
+    // MUST TEST WITH replica=1
+    constexpr uint64_t snp_lsn = 1;
+    constexpr uint64_t num_shards_per_pg = 3;
+    constexpr uint64_t num_batches_per_shard = 3;
+    constexpr uint64_t num_blobs_per_batch = 5;
+
+    // We have to create a PG first to init repl_dev
+    constexpr pg_id_t pg_id = 1;
+    create_pg(pg_id); // to create repl dev
+    PGStats stats;
+    ASSERT_TRUE(_obj_inst->pg_manager()->get_stats(pg_id, stats));
+    auto r_dev = homestore::HomeStore::instance()->repl_service().get_repl_dev(stats.replica_set_uuid);
+    ASSERT_TRUE(r_dev.hasValue());
+
+    auto handler = std::make_unique< homeobject::HSHomeObject::SnapshotReceiveHandler >(
+        *_obj_inst, stats.replica_set_uuid, r_dev.value());
+    handler->reset_context(snp_lsn, pg_id);
+
+    // Step 1: Test write pg meta - cannot test full logic since the PG already exists
+    // Generate ResyncPGMetaData message
+    LOGINFO("TESTING: applying meta for pg {}", pg_id);
+    constexpr auto blob_seq_num = num_shards_per_pg * num_batches_per_shard * num_blobs_per_batch;
+    flatbuffers::FlatBufferBuilder builder;
+    std::vector< flatbuffers::Offset< Member > > members;
+    std::vector uuid(stats.replica_set_uuid.begin(), stats.replica_set_uuid.end());
+    for (auto& member : stats.members) {
+        auto id = std::vector< std::uint8_t >(std::get< 0 >(member).begin(), std::get< 0 >(member).end());
+        members.push_back(CreateMemberDirect(builder, &id, std::get< 1 >(member).c_str(), 100));
+    }
+    std::vector< uint64_t > shard_ids;
+    for (uint64_t i = 1; i <= num_shards_per_pg; i++) {
+        shard_ids.push_back(i);
+    }
+    auto pg_entry =
+        CreateResyncPGMetaDataDirect(builder, pg_id, &uuid, blob_seq_num, num_shards_per_pg, &members, &shard_ids);
+    builder.Finish(pg_entry);
+    auto pg_meta = GetResyncPGMetaData(builder.GetBufferPointer());
+    builder.Reset();
+
+    handler->process_pg_snapshot_data(*pg_meta);
+
+    // Step 2: Test shard and blob batches
+    blob_id_t cur_blob_id{0};
+    for (uint64_t i = 1; i <= num_shards_per_pg; i++) {
+        LOGINFO("TESTING: applying meta for shard {}", i);
+
+        // Step 2-1: Test write shard meta
+        // Generate ResyncShardMetaData message
+        ShardInfo shard;
+        shard.id = i;
+        shard.state = ShardInfo::State::OPEN;
```

Comment:
Suggest covering the sealed shard and delete shard process. In the current implementation, handler will call `home_obj_.local_create_shard` for all shards no matter their states, but `local_create_shard` will remove the selected chunk from the heap. So seems handler should call seal_shard for sealed shard. Please correct me if I missed some logic.

Reply:
Thanks, nice catch. Fixed in latest commit.

Since currently it requires to change the method signature of `process_shard_snapshot_data` to know the allocated chunk ID, we'll finish the test after rebasing onto identical layout commits.